include/sys/zstd/zstd.h

```diff
@@ -94,6 +97,123 @@ int zfs_zstd_decompress(void *s_start, void *d_start, size_t s_len,
     size_t d_len, int n);
 void zfs_zstd_cache_reap_now(void);
 
+/*
+ * So, the reason we have all these complicated set/get functions is that
+ * originally, in the zstd "header" we wrote out to disk, we used a 32-bit
+ * bitfield to store the "level" (8 bits) and "version" (24 bits).
+ *
+ * Unfortunately, bitfields make few promises about how they're arranged in
+ * memory...
+ *
+ * As a consequence, we wound up with little endian platforms with a layout
+ * like this in memory:
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      | level |   0   |    version    |
+ *      +-------+-------+-------+-------+
+ *
+ * ...and then, after being run through BE_32(), serializing this out to
+ * disk (BS16 indicates an order swap between the two bytes):
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      | BS16(version) |   0   | level |
+ *      +-------+-------+-------+-------+
+ *
+ * while on big-endian systems, since BE_32() is a noop there, both in
+ * memory and on disk, we wind up with:
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      | level | BS16(version) |   0   |
```

Comment:
I don't understand why this is notated as "BS16(version)", because the bytes are not swapped. It's the same byte order in memory and on disk (big endian).  Are you using BS() to mean "big endian version of"?

Reply:
I annotated it that way because, if we're referring to "version" in the other two examples here as "v0 v1", then this is "v1 v0".

If that's not a swap, then I'm not sure what you'd describe it as.

Reply:
I didn't see the notation "v0 v1" in the comment, just "version", which I would assume is host-endian.

Reply:
Is the updated version clear enough?